lua/ui/game/reclaim.lua

```diff
@@ -1,497 +1,494 @@
+    ---@param position? Vector
+    __init = function(self, parent, icon, label, position)
+        Label.__init(self, parent, icon, label, 10, UIUtil.bodyFont, true)
 
-        self.text = UIUtil.CreateText(self, "", 10, UIUtil.bodyFont)
-        self.text:SetColor('ffc7ff8f')
-        self.text:SetDropShadow(true)
-        LayoutHelpers.AtLeftIn(self.text, self, 16)
-        LayoutHelpers.AtVerticalCenterIn(self.text, self)
+        self.parent = parent
+        self.position = position or Vector(0, 0, 0)
+        self.reproject = true
 
         self:DisableHitTest(true)
         self:SetNeedsFrameUpdate(true)
-    end,
-
-    --- Updates the reclaim that this label displays
-    ---@param self WorldLabel
-    ---@param r any
-    DisplayReclaim = function(self, r)
-        if self:IsHidden() then
-            self:Show()
+    end;
+
+    Layout = function(self)
+        Label.Layout(self)
+        local parent = self.parent
+        local icon, text = self.icon, self.text
+        -- Make all text show over icons
+        if icon then
+            LayoutHelpers.DepthOverParent(icon, parent, 3)
         end
-
-        self.position = r.position
-        if r.mass ~= self.oldMass then
-            local mass = tostring(math.floor(0.5 + r.mass))
-            self.text:SetText(mass)
-            self.oldMass = r.mass
+        if text then
+            LayoutHelpers.DepthOverParent(text, parent, 5)
         end
-    end,
+    end;
+
+    MoveTo = function(self, position)
+        self.position = position
+        self:PositionAt(position)
+    end;
+
+    PositionAt = function(self, position)
+        local proj = self.parent.view:Project(position)
+        self:FocusOn(proj[1], proj[2])
+    end;
 
     --- Called each frame by the engine
     ---@param self WorldLabel
     ---@param delta number
     OnFrame = function(self, delta)
-        if self._project then
-            local view = self.parent.view
-            local proj = view:Project(self.position)
-            self.Left:SetValue(proj.x - 0.5 * self.Width())
-            self.Top:SetValue(proj.y - 0.5 * self.Height() + 1)
+        if self.reproject then
+            self:PositionAt(self.position)
         end
-        self._project = self.parent.isMoving
-    end,
+        self.reproject = self.parent.isMoving
+    end;
 
     --- Called when the control is hidden or shown, used to start updating
     ---@param self WorldLabel
     ---@param hidden boolean
     OnHide = function(self, hidden)
         self:SetNeedsFrameUpdate(not hidden)
-    end,
+        if not hidden then
+            self.reproject = true
+        end
+    end;
 }
 
----
----@param a UIReclaimDataPoint
----@param b UIReclaimDataPoint
----@return boolean
-local function CompareMass(a, b)
-    return a.mass < b.mass
-end
 
----@type number
-local mapWidth = 0
+---@class ReclaimLabel : WorldLabel
+---@field mass number
+---@field count number
+local ReclaimLabel = Class(WorldLabel) {
```

Comment:
not sure what the difference is with the world label, given that the `WorldLabel` is a bad name 😄

Reply:
The idea was that `WorldLabel` is a general HUD marker anyone can use to bind a label to a world position--and `ReclaimLabel` is the label specific to the reclaim

Reply:
We shouldn't encourage translating world coordinates to screen coordinates. Using the project function is very, very bad for performance. We're just trying to skip it as much as possible here.